### ISSUE TITLE
Population does not work for Users in Users-Permissions

### ISSUE THREAD

**Issue by kamal-choudhary** [strapi/strapi#11957]
## Bug report

When we try to populate the relational fields on the user-permissions => users model, the response doesn't populate the relational fields.

### Steps to reproduce the behavior

1. Create a user in the Users-Permissions
2. Create some other collection type and link it with user
3. Enable public find permission on both
4. Send a GET request to the /users with "?populate=*"
5. Response won't contain the relational field

### Expected behavior

Appending ?populate=* to the API request to the user-permissions => users model, the response should contain the relational fields.

### System

- Node.js version: 14.15.3
- NPM version: 6.14.9
- Strapi version: 4.0.0
- Database: MariaDB
- Operating system: MacOS

### Additional context

PR  #11795 is about fixing filtering on the same User-permissions => users model. I have also added that new code changes to my project. Filtering does work, but the population doesn't.

**Comment by zorox-1** [comment]
facing same issue

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/populate-users-permissions-to-get-user-avatar/13701/2

**Comment by joh-sch** [comment]
We're having the exact same issue with getting user roles. No matter the population parameters, the user role is not in the response.

**Comment by Devisione** [comment]
to many issues with populate.... mb we can return default mechanism? or how we can help to resolve that problem?

**Comment by joh-sch** [comment]
> to many issues with populate.... mb we can return default mechanism? or how we can help to resolve that problem?

looks like a fix is on the way: [Fix unable to populate User in Users-Permissions](https://github.com/strapi/strapi/pull/11960)

**Comment by derrickmehaffy** [comment]
> to many issues with populate.... mb we can return default mechanism? or how we can help to resolve that problem?

No there is no way to return to the v3 format, we expected bugs in the first few releases.

**Comment by thomaslecoeur** [comment]
The issue also happens for the ```/me``` route.

**Comment by iicdii** [comment]
Temporary workaround to enable `populate` in `/api/users/me` and `/api/users`
```js
// path: src/extensions/users-permissions/strapi-server.js
module.exports = plugin => {
  const sanitizeOutput = (user) => {
    const {
      password, resetPasswordToken, confirmationToken, ...sanitizedUser
    } = user; // be careful, you need to omit other private attributes yourself
    return sanitizedUser;
  };

  plugin.controllers.user.me = async (ctx) => {
    if (!ctx.state.user) {
      return ctx.unauthorized();
    }
    const user = await strapi.entityService.findOne(
      'plugin::users-permissions.user',
      ctx.state.user.id,
      { populate: ['role'] }
    );

    ctx.body = sanitizeOutput(user);
  };

  plugin.controllers.user.find = async (ctx) => {
    const users = await strapi.entityService.findMany(
      'plugin::users-permissions.user',
      { ...ctx.params, populate: ['role'] }
    );

    ctx.body = users.map(user => sanitizeOutput(user));
  };

  return plugin;
};
```
I forced to populate `role` in example but you can edit it with your own.

**Comment by kevinvugts** [comment]
Adding a connecting an image to a users-permissions "user" model after creation also doesn't seem to work fokes.

After creating a user account and trying to upload the image to strapi and connect it to the users object only the upload seems to work. The coupling between upload and model is not happening for some reason.

Any ideas? @derrickmehaffy 

`        data.append("files", {
          uri: image,
          name: filename,
          type: `image/${filename.split(".").pop()}`,
        });
        data.append("ref", "user");
        data.append("refId", response.data.user.id);
        data.append("field", "profilePicture");
        data.append("source", "users-permissions");
`

**Comment by evilrat** [comment]
is there any workaround or fix for this issue.....still unable to populate users

**Comment by kevinvugts** [comment]
> #11919

Hi @evilrat,

There is an answer in this issue: #11919

Best

**Comment by evilrat** [comment]
> > #11919
> 
> Hi @evilrat,
> 
> There is an answer in this issue: #11919
> 
> Best

it's not what i am looking for.....i want to populate role after login 

api/auth/local?[populate]=*

**Comment by kevinvugts** [comment]
> > > #11919
> > 
> > 
> > Hi @evilrat,
> > There is an answer in this issue: #11919
> > Best
> 
> it's not what i am looking for.....i want to populate role after login
> 
> api/auth/local?[populate]=*

Hi EvilRat,

I Gotcha. Unfortunately the "populate" query string right now of speaking isn't working for the users-permission plugin. Therefore I created a gist for you that works for now. 

**Important notes:**

- Change the populate field with the name of your field that contains the information you want to be populated. You can also add an array ["*"] to populate all values.

Please let me know if this helped you.

https://pastebin.com/f7PtK1gU

**Comment by Tragio** [comment]
I think I'm facing the same issue. When I do a POST to /api/liked-contents with:
```json
{
  "data": {
    "user": 1,
    "event": 1
  }
}
```
The event is appearing but the user does not appear. I also tried with `"user": [1],` and other variations just to make sure.

![Screenshot 2022-01-22 at 09 21 03](https://user-images.githubusercontent.com/5109593/150633049-cc17bb8b-6fb8-4ba2-9b0e-db6fefbf8163.jpg)

**Comment by benjamincanac** [comment]
I had the same issue and found it easier to override the `fetchAuthenticatedUser` method (https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/services/user.js#L69) instead of the controller as the `me` route use the user already populated in the state: https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/controllers/user.js#L182, it will also prevent fetching the user twice.

The `ctx.state.user` is populated through the `fetchAuthenticatedUser` method that you can override in `src/index.js`:

```js
module.exports = {
  register ({ strapi }) {
    strapi.service('plugin::users-permissions.user').fetchAuthenticatedUser = (id) => {
      return strapi
        .query('plugin::users-permissions.user')
        .findOne({ where: { id }, populate: ['role', 'teams.avatar'] })
    }
  }
}
```

**Comment by Purematics** [comment]
For me the workaround of @iicdii is working. 

But there is a general problem of relating to User collection since v4. To give api-users a role is a central feature and should be merged as soon as possible.

Is there still really no solution?

**Comment by ssshammi** [comment]
@benjamincanac  that did not do anything, how to call the same for api/auth/local path i get the user details after login. Also what do i need to do to populate=* by default so i don't have to match the fields name in API calls. @iicdii solution works but not for api/auth/local

**Comment by ssshammi** [comment]
> Temporary workaround to enable `populate` in `/api/users/me` and `/api/users`
> 
> ```js
> // path: src/extensions/users-permissions/strapi-server.js
> module.exports = plugin => {
>   const sanitizeOutput = (user) => {
>     const {
>       password, resetPasswordToken, confirmationToken, ...sanitizedUser
>     } = user; // be careful, you need to omit other private attributes yourself
>     return sanitizedUser;
>   };
> 
>   plugin.controllers.user.me = async (ctx) => {
>     if (!ctx.state.user) {
>       return ctx.unauthorized();
>     }
>     const user = await strapi.entityService.findOne(
>       'plugin::users-permissions.user',
>       ctx.state.user.id,
>       { populate: ['role'] }
>     );
> 
>     ctx.body = sanitizeOutput(user);
>   };
> 
>   plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> 
>   return plugin;
> };
> ```
> 
> I forced to populate `role` in example but you can edit it with your own.

I modified this a bit 

`// path: src/extensions/users-permissions/strapi-server.js
module.exports = plugin => {
  const sanitizeOutput = (user) => {
    const {
      password, resetPasswordToken, confirmationToken, ...sanitizedUser
    } = user; // be careful, you need to omit other private attributes yourself
    return sanitizedUser;
  };

  plugin.controllers.user.me = async (ctx) => {
  if (!ctx.state.user) {
      return ctx.unauthorized();
    }
    const user = await strapi.entityService.findOne(
      'plugin::users-permissions.user',
      ctx.state.user.id,
      { populate: ['role','displaypic'] }
    );

    ctx.body = sanitizeOutput(user);
  };
  
  plugin.controllers.user.findOne = async (ctx) => {
   const { id } = ctx.params.id;
   console.log("params "+ctx.params.id);
    const user = await strapi.entityService.findOne(
      'plugin::users-permissions.user',
		ctx.params.id,
       {populate: ['role','displaypic'] }
    );

    ctx.body = sanitizeOutput(user);
  };

  plugin.controllers.user.find = async (ctx) => {
    const users = await strapi.entityService.findMany(
      'plugin::users-permissions.user',
      { ...ctx.params, populate: ['role','displaypic'] }
    );

    ctx.body = users.map(user => sanitizeOutput(user));
  };

  return plugin;
};`

but how to add to api/auth/local endpoint do i need to do something for plugin.users-permissions.jwt #11960 This is pending approval it's not in 4.1.0 ?

**Comment by devnm21** [comment]
The issue seems to be persistent with other content-types having relation to users as well.
Or, is it just me?

**Comment by kamal-choudhary** [comment]
@devnm21 for me, only the "/users" route doesn't support the population of relational fields. Other routes even if they have a relation with the **user** works fine for me.

**Comment by brice-noowu** [comment]
> plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role','displaypic'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };

With this workaround, API query filters are not working anymore, how could I make it populate + take into account query filters ?

**Comment by Felpat0** [comment]
> @devnm21 for me, only the "/users" route doesn't support the population of relational fields. Other routes even if they have a relation with the **user** works fine for me.

I also can populate **user** from other endpoints, but not **role** for some reason

**Comment by Neyunse** [comment]
Any news for this?

**Comment by Saad-ISAA** [comment]
> Any news for this? In the 4.1.3 this bug persists.

yeah. facing the same issue

**Comment by gideonibemerejr** [comment]
> Temporary workaround to enable `populate` in `/api/users/me` and `/api/users`
> 
> ```js
> // path: src/extensions/users-permissions/strapi-server.js
> module.exports = plugin => {
>   const sanitizeOutput = (user) => {
>     const {
>       password, resetPasswordToken, confirmationToken, ...sanitizedUser
>     } = user; // be careful, you need to omit other private attributes yourself
>     return sanitizedUser;
>   };
> 
>   plugin.controllers.user.me = async (ctx) => {
>     if (!ctx.state.user) {
>       return ctx.unauthorized();
>     }
>     const user = await strapi.entityService.findOne(
>       'plugin::users-permissions.user',
>       ctx.state.user.id,
>       { populate: ['role'] }
>     );
> 
>     ctx.body = sanitizeOutput(user);
>   };
> 
>   plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> 
>   return plugin;
> };
> ```
> 
> I forced to populate `role` in example but you can edit it with your own.

Be **VERY** careful using this. The query ends up being extremely expensive. My current app has over 14000 users, maybe a few thousand concurrently and crashed using this to differentiate users vs admins in the app via roles. Really looking forward to a fix for this...

**Comment by Nisthar** [comment]
i am guessing this issue is hard to fix. Its been 4 months.

**Comment by derrickmehaffy** [comment]
> i am guessing this issue is hard to fix. Its been 4 months.

Quite yes, the team currently doesn't have the time in the very immediate future to work on it but we are working with a few members of the community who are looking into it.

**Comment by Nikoxx99** [comment]
> i am guessing this issue is hard to fix. Its been 4 months.

Probably alot of considerations in this case. Because of a bad population method can increase the times in very big queries. The current workaround is very slow when when db has thounsand or millions of items.



> > i am guessing this issue is hard to fix. Its been 4 months.
> 
> Quite yes, the team currently doesn't have the time in the very immediate future to work on it but we are working with a few members of the community who are looking into it.

Thanks for comment about it. You're doing great with this version imo.

**Comment by derrickmehaffy** [comment]
> > i am guessing this issue is hard to fix. Its been 4 months.
> 
> Probably alot of considerations in this case. Because of a bad population method can increase the times in very big queries. The current workaround is very slow when when db has thounsand or millions of items.
> 
> > > i am guessing this issue is hard to fix. Its been 4 months.
> > 
> > 
> > Quite yes, the team currently doesn't have the time in the very immediate future to work on it but we are working with a few members of the community who are looking into it.
> 
> Thanks for comment about it. You're doing great with this version imo.

Our primry concern isn't performance (that is a concern absolutely) but instead it's security. This plugin is a critical cross for the vast majority of our users so we have to take things carefully and slowly.

**Comment by JonasJW** [comment]
Is there any way to get the `role` of the user object when logging in via the `auth/local` endpoint? I really don't want to call the `users/me` endpoint after logging in just to get the `role`of the user. This worked just fine in v3 and I can't understand why fundamental features like these just don't work in what should be an improvement to v3.

**Comment by Purematics** [comment]
@JonasJW Agree 100 percent. I have to make 2 requests, one for login and one for getting role (react frontend). It is working!
Of course it is not nice....
And I also do not understand how this could happen in V4. It is a central feature! Why is it so hard to solve? Where are "security concerns" ?. The JWT is holding the role and when hitting an endpoint and you decode it as usual...

**Comment by Saad-ISAA** [comment]
@Purematics  Though you are right, it shouldn't be a much big issue to resolve. But you need to understand a few things on the strapi core.
The Plugin API seperates logically the Admin and other APIs.
and since users-permissions is a plugin it doesnt directly fall under the same hood as the other REST apis of the collection types.
In v4 this was different, hence it can be time taking. secondly the strapi team is working on this with some of the community members to get this resolved as mentioned by @derrickmehaffy  earlier in the comments above.,

I am also one of the persons who raised the issue here, but lets wait for the good news for it.
Meanwhile instead of making 2 calls for the Role, you can extend the login method in your extensions.

**Comment by Saad-ISAA** [comment]
> > Meanwhile instead of making 2 calls for the Role, you can extend the login method in your extensions.
> 
> I have been trying to achieve this but failed at it. Can you give some tips how to do this? I tried [this](https://github.com/strapi/strapi/issues/11957#issuecomment-1028843791) solution but that doesn't seem to work, as mentioned in a comment above

What are you trying to achieve tell me in a little detail, and i will try to help you out.

**Comment by JonasJW** [comment]
> > > Meanwhile instead of making 2 calls for the Role, you can extend the login method in your extensions.
> > 
> > 
> > I have been trying to achieve this but failed at it. Can you give some tips how to do this? I tried [this](https://github.com/strapi/strapi/issues/11957#issuecomment-1028843791) solution but that doesn't seem to work, as mentioned in a comment above
> 
> What are you trying to achieve tell me in a little detail, and i will try to help you out.

Well first I tried to extend the `fetchAuthenticatedUser` method like [this](https://github.com/strapi/strapi/issues/11957#issuecomment-1028843791) but that doesn't seem to work. When checking the source code of that function [here](https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/services/user.js#L69) I can see, that it already populates the `role` but the role is still not included in the response. Afaik this is because in `strapi.query` populate is being ignored?


So what I have done now is overwrite the controller for `auth/local` similar to what the folks above me did with `users/me` or `users/find`. This is quite complex because there is quite a lot of logic in that function and I would rather not do it this way but that was the only way I could get it to work. This is the overwrite method of the `auth/local` method:

    const _ = require("lodash");
    const {
      validateCallbackBody,
    } = require("@strapi/plugin-users-permissions/server/controllers/validation/auth");
    const utils = require("@strapi/utils");
    const { getService } = require("@strapi/plugin-users-permissions/server/utils");

    const { sanitize } = utils;
    const { ApplicationError, ValidationError } = utils.errors;

    const populate = ["role", "profilbild"];

    const emailRegExp =
      /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    const sanitizeUser = (user, ctx) => {
      console.log("'''", user);
      const { auth } = ctx.state;
      const userSchema = strapi.getModel("plugin::users-permissions.user");

      return sanitize.contentAPI.output(user, userSchema, { auth });
    };

    /**
     * Overwriting user-permission controllers because they don't populate fields and population in request don't work for plugins
     */
    module.exports = (plugin) => {

      plugin.controllers.auth.callback = async (ctx) => {
        const provider = ctx.params.provider || "local";
        const params = ctx.request.body;

        const store = await strapi.store({
          type: "plugin",
          name: "users-permissions",
        });

        if (provider === "local") {
          if (!_.get(await store.get({ key: "grant" }), "email.enabled")) {
            throw new ApplicationError("This provider is disabled");
          }

          await validateCallbackBody(params);

          // Check if the provided identifier is an email or not.
          const isEmail = emailRegExp.test(params.identifier);

          const filter = { $and: [{ provider: provider}] };

          if (isEmail) {
            // query.email = params.identifier.toLowerCase();
            filter.$and.push({ email: params.identifier.toLowerCase() })
          } else {
            filter.$and.push({ email: params.identifier })
          }

          // Check if the user exists.
          const users = await strapi.entityService.findMany(
            "plugin::users-permissions.user",
            {
              filters: filter ,
              populate: populate,
            }
          );

          console.log(users);

          if (!users || users.length <= 0) {
            throw new ValidationError("Invalid identifier or password");
          }

          const user = users[0];

          if (
            _.get(await store.get({ key: "advanced" }), "email_confirmation") &&
            user.confirmed !== true
          ) {
            throw new ApplicationError("Your account email is not confirmed");
          }

          if (user.blocked === true) {
            throw new ApplicationError(
              "Your account has been blocked by an administrator"
            );
          }

          // The user never authenticated with the `local` provider.
          if (!user.password) {
            throw new ApplicationError(
              "This user never set a local password, please login with the provider used during account creation"
            );
          }

          const validPassword = await getService("user").validatePassword(
            params.password,
            user.password
          );

          if (!validPassword) {
            throw new ValidationError("Invalid identifier or password");
          } else {
            ctx.send({
              jwt: getService("jwt").issue({
                id: user.id,
              }),
              user: {
                ...(await sanitizeUser(user, ctx)),
                role: user.role,
                profilbild: user.profilbild,
              },
            });
          }
        } else {
          if (!_.get(await store.get({ key: "grant" }), [provider, "enabled"])) {
            throw new ApplicationError("This provider is disabled");
          }

          // Connect the user with the third-party provider.
          let user;
          let error;
          try {
            [user, error] = await getService("providers").connect(
              provider,
              ctx.query
            );
          } catch ([user, error]) {
            throw new ApplicationError(error.message);
          }

          if (!user) {
            throw new ApplicationError(error.message);
          }

          ctx.send({
            jwt: getService("jwt").issue({ id: user.id }),
            user: {
              ...(await sanitizeUser(user, ctx)),
              role: user.role,
              profilbild: user.profilbild,
            },
          });
        }
      };

      return plugin;
    };

Would appreciate your feedback, If I could to it more efficiently

**Comment by Purematics** [comment]
> @Purematics Though you are right, it shouldn't be a much big issue to resolve. But you need to understand a few things on the strapi core. The Plugin API seperates logically the Admin and other APIs. and since users-permissions is a plugin it doesnt directly fall under the same hood as the other REST apis of the collection types. In v4 this was different, hence it can be time taking. secondly the strapi team is working on this with some of the community members to get this resolved as mentioned by @derrickmehaffy earlier in the comments above.,
> 
> I am also one of the persons who raised the issue here, but lets wait for the good news for it. Meanwhile instead of making 2 calls for the Role, you can extend the login method in your extensions.

Ok thank you for clarifying.

**Comment by Purematics** [comment]
I think sanitize removes sensitive fields. This could be the reason that "roles" are marked as sensitive, thus will not shown up ?
For myself, code looks clean.

**Comment by JonasJW** [comment]
> I think sanitize removes sensitive fields. This could be the reason that "roles" are marked as sensitive, thus will not shown up ? For myself, code looks clean.

Ohh yes, you are right. The user also gets populated when doing `const user = await strapi.query('plugin::users-permissions.user').findOne({ where: query, populate: ['role'] },);`, so It would be enough to only overwrite the `sanitizeUser` method

**Comment by Purematics** [comment]
> > I think sanitize removes sensitive fields. This could be the reason that "roles" are marked as sensitive, thus will not shown up ? For myself, code looks clean.
> 
> Ohh yes, you are right. The user also gets populated when doing `const user = await strapi.query('plugin::users-permissions.user').findOne({ where: query, populate: ['role'] },);`, so It would be enough to only overwrite the `sanitizeUser` method

Yes, I think so. This is also a nice solution for my app.

**Comment by Nisthar** [comment]
> ```js
> plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role','displaypic'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> ```
> 
> With this workaround, API query filters are not working anymore, how could I make it populate + take into account query filters ?

can you use `if(ctx.query.populate)` and only populate if its specified?

**Comment by Saad-ISAA** [comment]
> > ```js
> > plugin.controllers.user.find = async (ctx) => {
> >     const users = await strapi.entityService.findMany(
> >       'plugin::users-permissions.user',
> >       { ...ctx.params, populate: ['role','displaypic'] }
> >     );
> > 
> >     ctx.body = users.map(user => sanitizeOutput(user));
> >   };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With this workaround, API query filters are not working anymore, how could I make it populate + take into account query filters ?
> 
> can you use `if(ctx.query.populate)` and only populate if its specified?

I actually was doing the same thing right now figuring out a way to use the query filters in entityAPI.
will let you know what i figure out..

**Comment by Nisthar** [comment]
> I actually was doing the same thing right now figuring out a way to use the query filters in entityAPI. will let you know what i figure out..

Sure that'd be great.

**Comment by OmarMtya** [comment]
Nothing changed? I'm looking to avoid using any custom controller solution

**Comment by HxX2** [comment]
I came up with this workaround with custom controllers overriding the existing ones.
just create a new file in src/extensions/user-permissions/ named strapi-server.js
with the following code: 
```js
let cleanOutput = (user) => {
    const {
        password, resetPasswordToken, confirmationToken, updatedBy, createdBy, ...sanitizedUser
    } = user;
    return sanitizedUser;
};


module.exports = plugin => {

  /**
   * Retrieve state user record.
   */

    plugin.controllers.user.me = async (ctx) => {
        if (!ctx.state.user) {
            return ctx.unauthorized();
        }
        const user = await strapi.entityService.findOne(
            'plugin::users-permissions.user',
            ctx.state.user.id, {
            populate: ctx.request.query.populate == '*' ? '*' : [ctx.request.query.populate == undefined ? '' : ctx.request.query.populate]
        }
        );

        ctx.body = cleanOutput(user);
    };

  /**
   * Retrieve users records.
   */

    plugin.controllers.user.find = async (ctx) => {

        if (!ctx.state.user) {
            return ctx.Forbidden('Forbidden')
        }

        const users = await strapi.entityService.findMany('plugin::users-permissions.user', {
            ...ctx.params,
            populate: ctx.request.query.populate == '*' ? '*' : [ctx.request.query.populate == undefined ? '' : ctx.request.query.populate]
        });

        ctx.body = users.map(user => cleanOutput(user));
    };

  /**
   * Retrieve a user record.
   */

    plugin.controllers.user.findOne = async (ctx) => {
        const { id } = ctx.params;

        const user = await strapi.entityService.findOne('plugin::users-permissions.user', id , {
            ...ctx.params,
            populate: ctx.request.query.populate == '*' ? '*' : [ctx.request.query.populate == undefined ? '' : ctx.request.query.populate]
        });

        ctx.body = cleanOutput(user);
    };
}
```

**Comment by cjboco** [comment]
> Temporary workaround to enable `populate` in `/api/users/me` and `/api/users`
> 
> ```js
> // path: src/extensions/users-permissions/strapi-server.js
> module.exports = plugin => {
>   const sanitizeOutput = (user) => {
>     const {
>       password, resetPasswordToken, confirmationToken, ...sanitizedUser
>     } = user; // be careful, you need to omit other private attributes yourself
>     return sanitizedUser;
>   };
> 
>   plugin.controllers.user.me = async (ctx) => {
>     if (!ctx.state.user) {
>       return ctx.unauthorized();
>     }
>     const user = await strapi.entityService.findOne(
>       'plugin::users-permissions.user',
>       ctx.state.user.id,
>       { populate: ['role'] }
>     );
> 
>     ctx.body = sanitizeOutput(user);
>   };
> 
>   plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> 
>   return plugin;
> };
> ```
> 
> I forced to populate `role` in example but you can edit it with your own.

This was the only thing that has worked for me so far. Thank you!

**Comment by violabg** [comment]
can't find any solution for graphql, using extensions/users-permissions/strapi-server.js, only works for REST, but is not call when calling usersPermissionsUsers in graphql, it looks that with V4 strapi is more oriented to the REST API part.
Working with graphql is getting frustrating.

**Comment by sashman-jaspin** [comment]
This is a really disappointing issue.

**Comment by Paiman-Rasoli** [comment]
I also have a Modal that has relation to users-permissions but when I request I don't get user information.
<img src="https://i.stack.imgur.com/3FYHr.png" alt="reponse-modal" />

result from the postman when I send the request.

<img src="https://i.stack.imgur.com/x50DD.png" alt="postman" />

**Comment by james-theretreater** [comment]
This bug is affecting us at the moment.

**Comment by Paiman-Rasoli** [comment]
I have an image field in the user model and I wanted to populate it after login so I could achieve it like this.
Now, after sending a request to `api/auth/local` you can see the avatar field is populated.☺
```javascript
// src/extensions/users-permissions/strapi-server.js

const _ = require("lodash");
const { validateCallbackBody } = require("@strapi/plugin-users-permissions/server/controllers/validation/auth");

const { getService } = require("@strapi/plugin-users-permissions/server/utils");
const { sanitize } = require("@strapi/utils");

module.exports = (plugin) => {
//sanitize-function
  const sanitizeUser = (user, ctx) => {
    const { auth } = ctx.state;
    const userSchema = strapi.getModel("plugin::users-permissions.user");
    return sanitize.contentAPI.output(user, userSchema, { auth });
  };
  // edit-callback-on-login
  plugin.controllers.auth.callback = async (ctx) => {
    const emailRegExp =
      /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    const provider = ctx.params.provider || "local";
    const params = ctx.request.body;
    const store = await strapi.store({
      type: "plugin",
      name: "users-permissions",
    });

    if (provider === "local") {
      if (!_.get(await store.get({ key: "grant" }), "email.enabled")) {
        throw new ApplicationError("This provider is disabled");
      }

      await validateCallbackBody(params);

      const query = { provider };
      console.log("KOKOKOK");

      // Check if the provided identifier is an email or not.
      const isEmail = emailRegExp.test(params.identifier);

      // Set the identifier to the appropriate query field.
      if (isEmail) {
        query.email = params.identifier.toLowerCase();
      } else {
        query.username = params.identifier;
      }

      // Check if the user exists.
      //populate any field you wnat in my case avatar is the field which I want to populate.
      const user = await strapi
        .query("plugin::users-permissions.user")
        .findOne({ where: query, populate: ["avatar"] });

      if (!user) {
        throw new ValidationError("Invalid identifier or password");
      }

      if (
        _.get(await store.get({ key: "advanced" }), "email_confirmation") &&
        user.confirmed !== true
      ) {
        throw new ApplicationError("Your account email is not confirmed");
      }

      if (user.blocked === true) {
        throw new ApplicationError(
          "Your account has been blocked by an administrator"
        );
      }

      // The user never authenticated with the `local` provider.
      if (!user.password) {
        throw new ApplicationError(
          "This user never set a local password, please login with the provider used during account creation"
        );
      }

      const validPassword = await getService("user").validatePassword(
        params.password,
        user.password
      );

      if (!validPassword) {
        throw new ValidationError("Invalid identifier or password");
      } else {
        console.log("Sanitized", await sanitizeUser(user, ctx));

        ctx.send({
          jwt: getService("jwt").issue({
            id: user.id,
          }),
          user: await sanitizeUser(user, ctx),
        });
      }
    } else {
      if (!_.get(await store.get({ key: "grant" }), [provider, "enabled"])) {
        throw new ApplicationError("This provider is disabled");
      }

      // Connect the user with the third-party provider.
      let user;
      let error;
      try {
        [user, error] = await getService("providers").connect(
          provider,
          ctx.query
        );
      } catch ([user, error]) {
        throw new ApplicationError(error.message);
      }

      if (!user) {
        throw new ApplicationError(error.message);
      }
      const explore = await sanitizeUser(user, ctx);
      ctx.send({
        jwt: getService("jwt").issue({ id: user.id }),
        user: explore,
      });
    }
    // end
  };

  return plugin;
}
```

**Comment by selected-pixel-jameson** [comment]
This is an extremely significant bug. I would suggest that Strapi call it out in a much more obvious way then just having it be a pinned issue on the Github repo. Maybe put a big red box on the documentation related to populating fields that this currently doesn't work for the User collection.  In fact it's quite the opposite. You have documentation stating that it is possible, which makes this even more confusing.

https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest/populating-fields.html#relation-media-fields
```
If the users-permissions plugin is installed, the find permission must be enabled for the content-types that are being populated. If a role doesn't have access to a content-type it will not be populated.
```

It's pretty frustrating when you are following documentation and things aren't working. I'm sure a lot of people, myself included, have spent a lot of time just banging their heads against the wall trying to figure out what they are doing wrong only to later stumble across this issue in the Github repo. 

This is a very poor experience at the moment. Things like this make me question my choice to use Strapi, as this has been flagged as a stable release for months at this point, yet you have significant issues and miscommunications like this. This issue has been open for 4 months. It's time to address it more formally than a pinned Github issue.

**Comment by james-theretreater** [comment]
I was able to solve this for the `me` query at least by extending `src/index.js` with this. Inspiration taken from https://github.com/strapi/strapi/issues/12016#issuecomment-1040378232

```
  register({ strapi }) {
    const extensionService = strapi.plugin("graphql").service("extension");
    extensionService.use(({ nexus }) => ({
      types: [
        nexus.extendType({
          type: "UsersPermissionsMe",
          definition(t) {
            t.field("role", {
              type: "String",
              resolve: async (root, args) => {
                const userData = await strapi.db
                  .query("plugin::users-permissions.user")
                  .findOne({
                    select: [],
                    where: { id: root.id },
                    populate: { role: true },
                  });
                return userData.role.type;
              },
            });
          },
        }),
      ],
    }));
  },
```

**Comment by james-theretreater** [comment]
> populate

Thank you! This works for me to get the `role` on the login response. It will do until this is fixed, I will have to manually check & update this file if it changes in future Strapi releases (except the one that fixes the populate role data).

**Comment by garfiaslopez** [comment]
> ```js
> ```js
> plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role','displaypic'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> With this workaround, API query filters are not working anymore, how could I make it populate + take into account query filters ?
> 


this is something i did that worked taking into account `filters` and `populate`.

```javascript 

  plugin.controllers.user.find = async (ctx) => {
    let config = {};
    const params = ctx.request.url.split('?');
    if (params.length >= 2) {
      config = qs.parse(params[1]);
    }
    const users = await strapi.entityService.findMany(
      'plugin::users-permissions.user',
      config
    );

    ctx.body = users.map(user => sanitizeOutput(user));
  };
  
  
```

**Comment by egornmore** [comment]
> Temporary workaround to enable `populate` in `/api/users/me` and `/api/users`
> 
> ```js
> // path: src/extensions/users-permissions/strapi-server.js
> module.exports = plugin => {
>   const sanitizeOutput = (user) => {
>     const {
>       password, resetPasswordToken, confirmationToken, ...sanitizedUser
>     } = user; // be careful, you need to omit other private attributes yourself
>     return sanitizedUser;
>   };
> 
>   plugin.controllers.user.me = async (ctx) => {
>     if (!ctx.state.user) {
>       return ctx.unauthorized();
>     }
>     const user = await strapi.entityService.findOne(
>       'plugin::users-permissions.user',
>       ctx.state.user.id,
>       { populate: ['role'] }
>     );
> 
>     ctx.body = sanitizeOutput(user);
>   };
> 
>   plugin.controllers.user.find = async (ctx) => {
>     const users = await strapi.entityService.findMany(
>       'plugin::users-permissions.user',
>       { ...ctx.params, populate: ['role'] }
>     );
> 
>     ctx.body = users.map(user => sanitizeOutput(user));
>   };
> 
>   return plugin;
> };
> ```
> 
> I forced to populate `role` in example but you can edit it with your own.

The solution works but the role is populated only after page reload.

**Comment by violabg** [comment]
it stil doesn't work on graphQl after updating to version 4.1.10

```gql
query getUsers {
  usersPermissionsUser(id: 1) {
      data {
        id
        attributes {
          firstName
          lastName
          blocked
          role{
            data{
              attributes {
                type
              }
            }
          }
        }
      }
    }
}
```

```gql
{
  "errors": [
    {
      "message": "Forbidden access",
      "extensions": {
        "error": {
          "name": "ForbiddenError",
          "message": "Forbidden access",
          "details": {}
        },
        "code": "FORBIDDEN"
      }
    }
  ],
  "data": {
    "usersPermissionsUser": {
      "data": {
        "id": "1",
        "attributes": {
          "firstName": "xxx",
          "lastName": "xxx",
          "blocked": false,
          "role": null
        }
      }
    }
  }
}

**Comment by csotiriou** [comment]
> 

That's certaily much easier, and it works for me, thank you.

**Comment by csotiriou** [comment]
> I had the same issue and found it easier to override the `fetchAuthenticatedUser` method (https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/services/user.js#L69) instead of the controller as the `me` route use the user already populated in the state: https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/controllers/user.js#L182, it will also prevent fetching the user twice.
> 
> The `ctx.state.user` is populated through the `fetchAuthenticatedUser` method that you can override in `src/index.js`:
> 
> ```js
> module.exports = {
>   register ({ strapi }) {
>     strapi.service('plugin::users-permissions.user').fetchAuthenticatedUser = (id) => {
>       return strapi
>         .query('plugin::users-permissions.user')
>         .findOne({ where: { id }, populate: ['role', 'teams.avatar'] })
>     }
>   }
> }
> ```

That's certainly much easier, and it works for me, @benjamincanac, thank you very much

**Comment by ssshammi** [comment]
Sadly this does not work, for localhost:1337/api/auth/local  and i have to make additional call for to check me is autorised with role.

**Comment by sjoukedv** [comment]
@derrickmehaffy how would we override the `callback` inside `auth.js` so that it can populate fields on login response? (this saves another request to the me endpoint)

It is retrieving the user it sends here https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/controllers/auth.js#L105. Where the `connect` function does not respect the query parameters, because it just finds by e-mail: https://github.com/strapi/strapi/blob/master/packages/plugins/users-permissions/server/services/providers.js#L62.

Maybe we can keep populate query on `/auth/<provider>` routes.

Can we re-open this?

**Comment by sjoukedv** [comment]
The PR only allows population on `/api/users/me` (which is good, because I had to override this function).

What I need is the login to response to return all the custom relations that belong to a user either by being able to override a function (not a whole file of the plugin) or by some setting/default. It already does so for fields (by default), but not relations. 

Currently, you authenticate yourself with `api/auth/...`, get a token and the user properties, e.g. `newsletterOptIn`, but for the relation, a `subscription` for example, I would have to immediately call `/api/users/me` afterwards to get a 'full user object' on the client-side.

**Comment by derrickmehaffy** [comment]
That's intended, you really shouldn't pollute a login request with additional information as that's the point of the me endpoint.

**Comment by jonashaefele** [comment]
Whatever happened to this? Why was the issue closed? Seems to still be super messy. Both migrations docs from v3 => v4 and the v4 docs don't accurately describe population itself or population of the users from users-permissions.

I'm in the process of migrating a fairly complex app from v3 to v4 and most things I could somehow figure out with lot of back and forth between the docs, the half-finished migration scripts (and their open PRs), and the forums. 

But it seems for users-permissions, there are only workarounds and hacks. 

the auth/local callback right now returns every single field, even custom-added JSON fields (on the user object), but doesn't populate the `roles` field, which is kinda essential for auth? 

From what I gather that should be remedied by making a second call to `api/users/me` with `populate=*` (or populate=role`) which feels like total overkill. We already pollute the login request with every single other field on the user model, while not populating the roles for efficiency's sake?

Maybe we have the standard `scope` parameter, like every other auth provider?

**Comment by derrickmehaffy** [comment]
> Whatever happened to this? Why was the issue closed? Seems to still be super messy. Both migrations docs from v3 => v4 and the v4 docs don't accurately describe population itself or population of the users from users-permissions.
> 
> I'm in the process of migrating a fairly complex app from v3 to v4 and most things I could somehow figure out with lot of back and forth between the docs, the half-finished migration scripts (and their open PRs), and the forums.
> 
> But it seems for users-permissions, there are only workarounds and hacks.
> 
> the auth/local callback right now returns every single field, even custom-added JSON fields (on the user object), but doesn't populate the `roles` field, which is kinda essential for auth?
> 
> From what I gather that should be remedied by making a second call to `api/users/me` with `populate=*` (or populate=role`) which feels like total overkill. We already pollute the login request with every single other field on the user model, while not populating the roles for efficiency's sake?
> 
> Maybe we have the standard `scope` parameter, like every other auth provider?

We patched population in I believe v4.2.0 or v4.2.2 but yes we don't allow population during the login request as that is specifically just for logging in and getting the JWT. Additional "profile fetching" should be done in a /me call.

**Comment by jonashaefele** [comment]
Thanks @derrickmehaffy 

Think I got super confused mostly because the API has a new format (with the `id` and `attributes` separate), whereas the users-permissions plugin still follows the old data format (id and attributes in the same object). So population works differently there, too. That made it even harder to figure out how the population queries really work. Figuring it out now.

Definitely doing the separate profile fetch, that makes sense and is best practice. Thanks for pushing that.

But since we're already doing a second profile fetch where we specify which data we need, wouldn't it make sense to only return minimal data on the auth callback? Currently Strapi returns all fields on the user object, including large JSON fields in my custom model. Can we have a minimal auth callback response? Someting like username, email, and token details? 

And then fetch the profile data we actually need in a separate API call, as you mentioned. That would be super clean. Currently we're a half way between two approaches.

**Comment by derrickmehaffy** [comment]
Agreed it's a bit odd right now because we only updated users-permissions enough to get it functional on v4. Ideally in the future we would like to replace this package with a new one that will have a more features but the code is so old in U&P and we wanted to minimize breaking changes as much as we could in v4.

It was a bit of an oversight on our part that it's response format wasn't uniform (more info here: https://github.com/strapi/strapi/issues/12280 )

I'll reopen this as a discussion and try to get @marcoautiero or @alexandrebodin opinion on if we could tweak the auth response. My only fear is that since it has been like this since v4 launched it would be a breaking change. Maybe we could introduce a simple plugin config to select the default fields for the U&P (maybe population too) as a temporary option until we figure out plans to replace this package.

**Comment by derrickmehaffy** [comment]
Actually on second thought, could you open a new issue for this discussion as this issue is largely about population and I don't want to confuse others.

**Comment by Sefibrah** [comment]
I know that I am late to the party, but I've found the solution, **but** it involves messing around with the **package files** in 
```node_modules/@strapi/plugin-users-permissions```. I am not sure if this is even permissible, because for example if I were to host this on a some kind of hosting provider, I guess that I'll then need to modify their package file(s) as well. Please tell me if this is a hack or an obvious _no touch zone_:

```
// this is in node_modules/@strapi/plugin-users-permissions/server/controllers/user.js
// for example, go to the async me() function and update it with:

// ...
async me(ctx) {
    let user = ctx.state.user;
    // user.role = user.role.name;

    if (!user) {
      return ctx.unauthorized();
    }

    console.log('old ctx.body with no role', ctx.body);
    ctx.body = await sanitizeOutput(user, ctx);
    // my customisation
    ctx.body = {
      ...ctx.body,
      role: user.role
    }
    console.log('ctx.body with the user role!', ctx.body);
  },

// ...
```

This is what I got in Postman:

<img width="778" alt="Screenshot 2022-12-15 at 13 44 18" src="https://user-images.githubusercontent.com/36334146/207839339-4b95501d-efb5-4bfc-9b56-cb319843ee6f.png">

I suppose that you could do the same to find() and findOne() methods, and that's to modify the ctx.body.

**Comment by derrickmehaffy** [comment]
> **update:** this comment is pointless, everything will be overwritten as soon as you enter ```npm i```, but I've seen something about forking, maybe this can be useful for forks?
> 
> I know that I am late to the party, but I've found the solution, **but** it involves messing around with the **package files** in 
> ```node_modules/@strapi/plugin-users-permissions```. I am not sure if this is even permissible, because for example if I were to host this on a some kind of hosting provider, I guess that I'll then need to modify their package file(s) as well. Please tell me if this is a hack or an obvious _no touch zone_:
> 
> ```
> // this is in node_modules/@strapi/plugin-users-permissions/server/controllers/user.js
> // for example, go to the async me() function and update it with:
> 
> // ...
> async me(ctx) {
>     let user = ctx.state.user;
>     // user.role = user.role.name;
> 
>     if (!user) {
>       return ctx.unauthorized();
>     }
> 
>     console.log('old ctx.body with no role', ctx.body);
>     ctx.body = await sanitizeOutput(user, ctx);
>     // my customisation
>     ctx.body = {
>       ...ctx.body,
>       role: user.role
>     }
>     console.log('ctx.body with the user role!', ctx.body);
>   },
> 
> // ...
> ```
> 
> This is what I got in Postman:
> 
> <img width="778" alt="Screenshot 2022-12-15 at 13 44 18" src="https://user-images.githubusercontent.com/36334146/207839339-4b95501d-efb5-4bfc-9b56-cb319843ee6f.png">
> 
> I suppose that you could do the same to find() and findOne() methods, and that's to modify the ctx.body.
> 
> 

You could use patch-package to make this permanent.

**Comment by myesn** [comment]
I don't know why this issue is closed, getting `role` in the `/api/users/me` interface is a normal thing to do, getting `role` information via `?populate=role` doesn't pollute the user data at all because it's an additional interface parameter that doesn't return `role` information by default, what's wrong with that? I'm not sure why Strapi v4.8.2 still can't solve this problem, if you can't get the role information, how does the client go about making the permissions determination, based on the comments above, I learned that you need to write some code to override the default behavior to get the `role` information, which is just terrible.

![image](https://user-images.githubusercontent.com/18598579/227832968-d062ecf0-3c36-49a8-a7a3-03af6962198d.png)

**Comment by cpres** [comment]
Not working for me either! Haven't tried this yet but it could solve:
https://market.strapi.io/plugins/strapi-plugin-populate-deep

**Comment by mo-elnady** [comment]
> This issue has been mentioned on **Strapi Community Forum**. There might be relevant details there:
> 
> https://forum.strapi.io/t/mastering-flutter-authentication-with-strapi-cms-a-step-by-step-guide/33851/5

Yes, this works. But what's the point? now any registered user with his token can just fetch the rest of the users data as well, not just his!? doesn't seem secure to me!? 

Or am I missing something here?
